subsys/net/l2/ethernet/ethernet.c

```diff
@@ -774,8 +836,28 @@ enum net_l2_flags ethernet_flags(struct net_if *iface)
 	return ctx->ethernet_l2_flags;
 }
 
+static int ethernet_l2_alloc(struct net_if *iface, struct net_pkt *pkt,
+			     size_t size, enum net_ip_protocol proto,
+			     k_timeout_t timeout)
+{
+	size_t reserve_ll_header;
+	int ret;
+
+	if (!IS_ENABLED(CONFIG_NET_L2_ETHERNET_RESERVE_HEADER)) {
+		ret = -ENOTSUP;
+		goto out;
+	}
+
+	reserve_ll_header = get_reserve_ll_header_size(iface);
```

Comment:
your dedicated allocator knows the reserver size obviously, so after allocating the net_pkt, it could run net_buf_reserve() by itself.

Way less changes needed in net_pkt

Reply:
Problem is that now every allocator needs to do the magic things like calling `net_buf_reserve()`, init cursor etc. So we are duplicating the functionality to multiple places which increases chances that something goes wrong. For Ethernet, the reserve setting would need to be done in two places, in ethernet.c and in virtual.c for VLAN. This might look like a small thing, but if some related thing need to be fixed, one would need to remember to fix it in multiple places. And this would be difficult to do for out of tree code, which does seem quite unlikely atm but still possible.
To me it really looks like it would be better if we concentrate the reserve allocation / related initialization activities to net_pkt.c.
@tbursztyka WDYT?

Reply:
Looks like we are now with a list of pros/cons on a few things:

Generalizing the reserve into net_pkt indeed has the advantage to centralize all about allocation.
However, it unilaterally bloats net_pkt even for systems that will never need it. (note also this fix is for Ethernet only so far, question is: do we foresee any other l2 layer that could take advantage of this?)

On one hand, net_pkt_*_reserve is after all like net_buf_reserve, it brings a generic concept about reservation without the l2 one which is far better that earlier proposal.
On the other hand, it could be fully hidden from net_pkt public API as I proposed.

One fact to add: if we generalize it but only ethernet will use it, it will be more annoying to specialize it back - if ever required one day. The reverse is not an issue.

I am withdrawing my lock on this PR. I let other reviewers (@pdgendt @rlubos @krish2718) to make the final decision.